blog/2024-01-24-transforming-your-data-1/index.md

```diff
@@ -0,0 +1,471 @@
+
+We can do the same for the other two edge tables as well.
+
+```sql
+// Generate `belongs_to.csv` from `merchant.csv`
+COPY (
+    LOAD FROM 'merchant.csv' (header=true)
+    RETURN
+        merchant_id,
+        company_id
+)
+TO 'belongs_to.csv';
+```
+
+```sql
+// Generate `located_in.csv` from `merchant.csv`
+COPY (
+    LOAD FROM 'merchant.csv' (header=true)
+    RETURN
+        merchant_id,
+        city_id
+)
+TO 'located_in.csv';
+```
+
+With all the input files in place, we can now proceed to insert the data and build the graph!
+
+## Insert data into Kùzu
+
+Collecting all the above functions, we can write a script that performs the following:
+
+1. Transform node/edge tables as needed
+1. Create the node tables
+1. Create edge tables
+1. Load the node tables into the database
+1. Load the edge tables into the database
+
+In Python, it would look something like this:
+
+```sql
+// Copy from CSV to node tables
+COPY Client FROM 'client.csv' (header=true)
+COPY City FROM 'city.csv' (header=true)
+COPY Company FROM 'company.csv' (header=true)
+COPY Merchant FROM 'merchant.csv' (header=true)
+
+// Copy from CSV to edge tables
+COPY TransactedWith FROM 'transacted_with.csv' 
+COPY BelongsTo FROM 'belongs_to.csv'
+COPY LocatedIn FROM 'located_in.csv'
+```
+
+The queries above require that the empty tables were created beforehand. The `COPY <edge_table> FROM <file>` statement
+writes the data into a Kùzu database. Running the queries on an existing database connection
+results in the graph being saved to a local directory.
+
+## Querying the graph
+
+We then run some simple queries to test that the data was loaded correctly. We can either create
+a standalone script using the client SDK of your choice, or fire up a [Kùzu CLI](https://kuzudb.com/docusaurus/getting-started/cli)
+shell and run some Cypher queries.
+
+The first query finds all the clients who transacted with the merchants of "Starbucks".
+
+```sql
+// Q1. Who are the clients that transacted with the merchants of "Starbucks"?
+MATCH (c:Client)-[:TransactedWith]->(:Merchant)-[:BelongsTo]->(co:Company)
+WHERE co.company = "Starbucks"
+RETURN DISTINCT c.client_id AS id, c.name AS name;
+```
+
+id | name
+:---: | :---:
+2 | Boris
+4 | Diana
+5  | Eve
+
+Query 2 traverses multiple paths to find clients who transacted with at least 2 separate merchants.
+
+```sql
+// Q2. Who are the clients who transacted with at least 2 separate merchants operating in Los Angeles?
+MATCH (c:Client)-[:TransactedWith]->(m1:Merchant)-[:LocatedIn]->(ci:City),
+      (c)-[:TransactedWith]->(m2:Merchant)-[:LocatedIn]->(ci)
+WHERE ci.city = "Los Angeles" AND m1.merchant_id <> m2.merchant_id
+RETURN DISTINCT c.client_id AS id, c.name as name;
+```
+
+id | name
+:---: | :---:
+3 | Cecil
+
+We can study multiple paths involving m-n joins as follows.
+
+```sql
+// Q3. Which companies have merchants in New York City, Boston **and** Los Angeles?
```

Comment:
Is there a better way to express this sort of AND join? I just used multiple match statements.